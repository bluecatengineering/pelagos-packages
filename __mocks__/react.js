module.exports = {
	...jest.requireActual('react'),
	useState: jest.fn((v) => [typeof v === 'function' ? v() : v, jest.fn()]),
	useCallback: (fn) => fn,
	useContext: jest.fn(),
	useEffect: jest.fn(),
	useLayoutEffect: jest.fn(),
	useMemo: jest.fn((f) => f()),
	useReducer: jest.fn((reducer, initial) => [initial, jest.fn()]),
	useRef: jest.fn((current) => ({current})),
	Children: {
		toArray(children) {
			return Array.isArray(children) ? children : Array.of(children);
		},
		map(children, fn) {
			return this.toArray(children).map(fn);
		},
	},
};
