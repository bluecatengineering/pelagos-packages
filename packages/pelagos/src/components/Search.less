@import '../../less/spacing';
@import '../../less/utils';

@layer pelagos {
	.Search {
		position: relative;
		flex-direction: row;

		&__input[type='text'] {
			flex: 1;
			width: 0;
			padding: 0 @sp-48;
		}

		&__icon {
			position: absolute;
			top: @sp-08;
			left: @sp-16;
			color: var(--icon-secondary);
			pointer-events: none;
		}

		&__clear {
			@icon-button();
			position: absolute;
			top: 0;
			right: 0;
			width: 32px;
			height: 32px;
			border-bottom: 1px solid transparent;
			outline: 2px solid transparent;
			outline-offset: -2px;

			&::before {
				content: '';
				position: absolute;
				top: 2px;
				bottom: 1px;
				left: 0;
				width: 2px;
				background-color: var(--field);
				transition: background-color 0.15s ease-out;
			}

			&:hover {
				border-color: var(--border-strong);

				&::before {
					background-color: var(--field-hover);
				}
			}

			&:focus-visible {
				&::before {
					background-color: var(--focus);
				}
			}

			&:active {
				&::before {
					background-color: var(--layer-active);
				}
			}

			.Search__input:hover + &::before {
				background-color: var(--field-hover);
			}

			.Search__input:focus + &:hover {
				@focus-ring();
			}
		}

		&--disabled svg {
			color: var(--icon-disabled);
		}
	}
}
