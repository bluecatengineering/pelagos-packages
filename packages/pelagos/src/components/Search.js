import {useCallback, useRef} from 'react';
import PropTypes from 'prop-types';
import {t} from '@bluecateng/l10n.macro';
import Close from '@carbon/icons-react/es/Close';
import SearchIcon from '@carbon/icons-react/es/Search';

import Layer from './Layer';

import './Search.less';

/** Search input. */
export const Search = ({className, value, disabled, onChange, ...props}) => {
	const inputRef = useRef(null);
	const handleChange = useCallback((event) => onChange(event.target.value.toLowerCase()), [onChange]);
	const handleKeyDown = useCallback(
		(event) => {
			if (event.key === 'Escape') {
				onChange('');
			}
		},
		[onChange]
	);
	const handleClearClick = useCallback(() => (inputRef.current.focus(), onChange('')), [onChange]);
	return (
		<div className={`Search${disabled ? ' Search--disabled' : ''}${className ? ` ${className}` : ''}`}>
			<SearchIcon className="Search__icon" />
			<Layer
				{...props}
				as="input"
				className="Search__input"
				type="text"
				role="searchbox"
				value={value}
				disabled={disabled}
				ref={inputRef}
				onChange={handleChange}
				onKeyDown={handleKeyDown}
			/>
			{value && (
				<button className="Search__clear" disabled={disabled} aria-label={t`Clear search`} onClick={handleClearClick}>
					<Close />
				</button>
			)}
		</div>
	);
};

Search.propTypes = {
	/** The component class name(s). */
	className: PropTypes.string,
	/** The search text. Must be in lowercase. */
	value: PropTypes.string,
	/** Whether the field is disabled. */
	disabled: PropTypes.bool,
	/** Function invoked when the text changes. */
	onChange: PropTypes.func.isRequired,
};

export default Search;
