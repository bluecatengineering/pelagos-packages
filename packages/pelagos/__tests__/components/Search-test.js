import {useRef} from 'react';
import {shallow} from 'enzyme';

import Search from '../../src/components/Search';

jest.unmock('../../src/components/Search');

describe('Search', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(<Search id="test" value="test" onChange={jest.fn()} />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when optional properties are set', () => {
			const wrapper = shallow(<Search id="test" className="TestClass" value="test" disabled onChange={jest.fn()} />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('calls onChange when the text changes', () => {
			const onChange = jest.fn();
			const wrapper = shallow(<Search value="test" onChange={onChange} />);
			wrapper.find('input').simulate('change', {target: {value: 'TEST'}});
			expect(onChange.mock.calls).toEqual([['test']]);
		});

		it('calls onChange when escape key is pressed', () => {
			const onChange = jest.fn();
			const wrapper = shallow(<Search value="test" onChange={onChange} />);
			wrapper.find('input').simulate('keydown', {keyCode: 27});
			expect(onChange.mock.calls).toEqual([['']]);
		});

		it('does not call onChange when other key is pressed', () => {
			const onChange = jest.fn();
			const wrapper = shallow(<Search value="test" onChange={onChange} />);
			wrapper.find('input').simulate('keydown', {keyCode: 13});
			expect(onChange.mock.calls).toEqual([]);
		});

		it('calls onChange when the clear button is clicked', () => {
			const onChange = jest.fn();
			const focus = jest.fn();
			useRef.mockReturnValue({current: {focus}});
			const wrapper = shallow(<Search value="test" onChange={onChange} />);
			wrapper.find('button').simulate('click');
			expect(focus.mock.calls).toEqual([[]]);
			expect(onChange.mock.calls).toEqual([['']]);
		});
	});
});
