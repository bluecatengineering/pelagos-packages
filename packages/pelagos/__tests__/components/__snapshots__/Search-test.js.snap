// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search rendering renders expected elements 1`] = `
<div
  className="Search"
>
  <Search
    className="Search__icon"
  />
  <Layer
    as="input"
    className="Search__input"
    id="test"
    onChange={[Function]}
    onKeyDown={[Function]}
    role="searchbox"
    type="text"
    value="test"
  />
  <button
    aria-label="Clear search"
    className="Search__clear"
    onClick={[Function]}
  >
    <Close />
  </button>
</div>
`;

exports[`Search rendering renders expected elements when optional properties are set 1`] = `
<div
  className="Search Search--disabled TestClass"
>
  <Search
    className="Search__icon"
  />
  <Layer
    as="input"
    className="Search__input"
    disabled={true}
    id="test"
    onChange={[Function]}
    onKeyDown={[Function]}
    role="searchbox"
    type="text"
    value="test"
  />
  <button
    aria-label="Clear search"
    className="Search__clear"
    disabled={true}
    onClick={[Function]}
  >
    <Close />
  </button>
</div>
`;
